File: P21 Merge Two Sorted Lists/Solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    // Approach 1: Recursion
    // Time complexity: O(n + m)
    // Space complexity: O(n + m)
    public ListNode mergeTwoLists(ListNode l1, ListNode l2) {
        if (list1 == null) {
            return list2;
        } else if (list2 == null) {
            return list1;
        } else {
            ListNode mergedList;
            if (list1.val < list2.val) {
                mergedList = list1;
                mergedList.next = mergeTwoLists(list1.next, list2);
            } else {
                mergedList = list2;
                mergedList.next = mergeTwoLists(list1, list2.next);
            }
            return mergedList;
        }
    }

    // Approach 2: Iteration
    // Time complexity: O(n + m)
    // Space complexity: O(1)
    public ListNode mergeTwoLists2(ListNode l1, ListNode l2) {
        ListNode resultNode = new ListNode(-1);
        ListNode head = resultNode;
        while(list1 != null && list2 != null) {
            if(list1.val <= list2.val) {
                head.next = list1;
                list1 = list1.next;
            } else {
                head.next = list2;
                list2 = list2.next;
            }
            head = head.next;
        }

        head.next = (list1 != null) ? list1 : list2;
        ListNode temp = resultNode;
        resultNode = resultNode.next;
        return resultNode;
    }
}

File: P21 Merge Two Sorted Lists/Solution.kt
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    // Approach 1: Recursion
    // Time complexity: O(n + m)
    // Space complexity: O(n + m)
    fun mergeTwoLists(list1: ListNode?, list2: ListNode?): ListNode? {
        val mergedList: ListNode?
        when {
            list1 == null -> mergedList = list2
            list2 == null -> mergedList = list1
            else -> {
                if (list1.`val` < list2.`val`) {
                    mergedList = list1
                    mergedList.next = mergeTwoLists(list1.next, list2)
                } else {
                    mergedList = list2
                    mergedList.next = mergeTwoLists(list1, list2.next)
                }
            }
        }
        return mergedList
    }

    // Approach 2: Iteration
    // Time complexity: O(n + m)
    // Space complexity: O(1)
    fun mergeTwoLists2(list1: ListNode?, list2: ListNode?): ListNode? {
        val head = ListNode(-1)
        var tail: ListNode? = head
        var mutableList1 = list1
        var mutableList2 = list2
        while(mutableList1 != null && mutableList2 != null) {
            if (mutableList1.`val` < mutableList2.`val`) {
                tail?.next = mutableList1
                mutableList1 = mutableList1.next
            } else {
                tail?.next = mutableList2
                mutableList2 = mutableList2.next
            }
            tail = tail?.next
        }

        tail?.next = mutableList1 ?: mutableList2
        return head.next
    }
}